### ISSUE TITLE
mixup of multiplication and exponentiation in z3 translation

### ISSUE THREAD

**Issue by Nordwald** [bdcht/amoco#72]
I have a mean little assembler program to test cfg recovery of indirect branches. The gcc-compatible code is the following:

``` javascript
.global main
.intel_syntax noprefix

.extern getchar
.extern printf

.section .data
jmpTable:
    .long _stub0
    .long _stub1
    .long _stub2
fmt: .asciz "%x\n"


.section .text

main:
    call getchar
    mov dl, 4
    imul dl
    add eax, offset jmpTable
    jmp [eax]
    .long 3851
_stub0:
    mov eax, 0
    jmp end
    .long 3851
_stub1:
    mov eax, 1
    jmp end
    .long 3851
_stub2:
    mov eax, 2
    jmp end
    .long 3851
end:
    push eax
    push offset fmt
    call printf
    add esp, 8
    ret
```

A minimal example causing the Exception is the following:

``` python
import amoco
from amoco.cas.smt import solver

p = amoco.system.loader.load_program('example_indirect.o')

analysis = amoco.main.lbackward(p)
cfg = analysis.getcfg()

for node in cfg.V():
    print '-' * 20, node.data.address, '-' * 20
    print node.data
    print node.data.map
    for n in node.data.map:
        print n[0], n[1].to_smtlib()
```

The crash message is the following:

```
Traceback (most recent call last):
  File "amoco_test.py", line 18, in <module>
    print n[0], n[1].to_smtlib()
  File "amoco-release/amoco/cas/smt.py", line 78, in comp_to_z3
    parts = [x.to_smtlib(slv) for x in e]
  File "amoco-release/amoco/cas/smt.py", line 95, in mem_to_z3
    p = e.a.to_smtlib(slv)
  File "amoco-release/amoco/cas/smt.py", line 90, in ptr_to_z3
    return e.base.to_smtlib(slv)+e.disp
  File "amoco-release/amoco/cas/smt.py", line 78, in comp_to_z3
    parts = [x.to_smtlib(slv) for x in e]
  File "amoco-release/amoco/cas/smt.py", line 86, in slc_to_z3
    x = e.x.to_smtlib(slv)
  File "amoco-release/amoco/cas/smt.py", line 159, in op_to_z3
    res = op(z3l,z3r)
  File "amoco-release/amoco/cas/expressions.py", line 1402, in __call__
    return self.impl(l,r)
TypeError: unsupported operand type(s) for ** or pow(): 'instance' and 'instance'
```

The expression which should be translated is the following:

```
eip <- { | [0:32]->M32({ | [0:8]->(al**0x4)[0:8] | [8:16]->(al**0x4)[8:16] | [16:32]->eax[16:32] | }+134520864) | }
```

There seems to be a mixup of multiplication and exponentiation

**Comment by bdcht** [comment]
Hi, do you confirm that you are using the last committed source? this seems
related to a bug fixed recently. (the pow operator is used by amoco to
distinguish the long multiply from the "usual" multiply, so there is no
exponentiation involved here)

Thanks
Axel

**Comment by Nordwald** [comment]
Hi there,

just confirmed it using the latest GitHub code.

**Comment by bdcht** [comment]
Should be fixed by commit 33a3ec6. 
Long multiply support was missing from op_to_z3 translation method.
Thanks for reporting the issue.
